android/src/main/java/com/ryanheise/audioservice/AudioService.java

```diff
@@ -202,7 +203,16 @@ void setState(List<NotificationCompat.Action> actions, int actionBits, int[] com
 				.setState(playbackState, position, speed, updateTime);
 		mediaSession.setPlaybackState(stateBuilder.build());
 
-		updateNotification();
+		if(playbackState==PlaybackStateCompat.STATE_PAUSED) stopForegroundOnPause();
+		if (androidStopForegroundOnPause && !isForeground && playbackState == PlaybackStateCompat.STATE_PLAYING)
+			mediaSessionCallback.play(new Runnable() {
```

Comment:
It would be interesting to test whether this `play()` wrapper, which inserts some code before and after your runnable, must preserve this sequence, or whether it would work just as well if all steps were done before the runnable. That would certainly make it easier to call this logic as needed. And then I think the state mechanics (i.e. here) could be the only place it needs to be called. I am not yet sure on the argument for why this wouldn't work, assuming the plugin user ensures that they always call `setState` as needed.

Note that as this is, the above call could end up being redundant. That is, `play` is being called here, which creates the wakelock amongst other things, but that is *also* being done in `onPlay`, so you'll end up with this routine being called double in several places. Looking at the code, it has some guards in place to ensure that if it is called from a state where the wakelock already exists, etc. it will not create a second wakelock, but it seems cleaner to just not have these double calls in the first place.

What I'm proposing is to get rid of `play` altogether, and just have a routine that sets up the wakelock and all the other stuff (e.g. `startService`) and have that only called in `setState`. This change does make me more nervous than any other change, because it's a significant shift in how the plugin works, but I was never happy with the old approach and had always wanted to some day integrate it into the state mechanics "if possible".

Do you see any reason why this might not work?

Reply:
Current setup works and making these changes seems hard at the moment. Let the double calls be there for the moment.